### PR TITLE
[#2127] Derive `namespace` from transformation selected-column payload

### PR DESCRIPTION
**Pull request by tangrammer**
- [ ] **Update release notes if necessary**

to be merged on issue-2858 branch

### CODE REVIEW

File: backend/specs/akvo/lumen/specs/transformation.clj [hunk 1]
```diff
@@ -31,13 +31,10 @@
 (create-ns 'akvo.lumen.specs.transformation.engine)
 (alias 'transformation.engine.s 'akvo.lumen.specs.transformation.engine)
 
-(s/def ::transformation.engine.s/created (s/with-gen #(instance? Instant %) (fn [] lumen.s/instant-gen)))
+(s/def ::transformation.engine.s/created (s/with-gen #(instance? Instant (Instant/parse %)) (fn []  lumen.s/str-instant-gen)))
```
Comment (tangrammer) [inline backend/specs/akvo/lumen/specs/transformation.clj:34]: `created` is now a string instant representation

File: backend/specs/akvo/lumen/specs/transformation.clj [hunk 2]
```diff
@@ -31,13 +31,10 @@
...
 
 (s/def ::transformation.engine.s/onError #{"leave-empty" "fail" "delete-row" "default-value"})
 
-(s/def ::transformation.engine.s/namespace (s/with-gen string? #(s/gen #{"main"})))
```
Comment (tangrammer) [inline backend/specs/akvo/lumen/specs/transformation.clj:38]: we don't expect so far to have `namespace` in transformation payload

File: backend/src/akvo/lumen/db/dataset.clj [hunk 3]
```diff
@@ -22,24 +22,6 @@
         (assoc "groupId" groupId)
         (assoc "groupName" groupName))))
 
-(defn dataset-in-groups-by-id
```
Comment (tangrammer) [inline backend/src/akvo/lumen/db/dataset.clj:25]: remove death code

File: backend/src/akvo/lumen/lib/transformation.clj [hunk 4]
```diff
@@ -78,15 +79,16 @@
 
 (defn apply
   [{:keys [tenant-conn] :as deps} dataset-id command]
-  (if-let [current-tx-job (db.transformation/pending-tx-or-update-job-execution tenant-conn {:dataset-id dataset-id})]
```
Comment (tangrammer) [inline backend/src/akvo/lumen/lib/transformation.clj:81]: this change only adds `created` to the transformation payload

File: backend/src/akvo/lumen/lib/transformation/change_datatype.clj [hunk 5]
```diff
@@ -101,7 +101,8 @@
   [{:keys [tenant-conn]} dataset-versions op-spec]
   (let [{column-name "columnName"
          new-type "newType"} (engine/args op-spec)
-        namespace (engine/get-namespace op-spec)
```
Comment (tangrammer) [inline backend/src/akvo/lumen/lib/transformation/change_datatype.clj:104]: we now derive `namespace` from transformation selected-column payload

File: backend/src/akvo/lumen/lib/transformation/combine.clj [hunk 6]
```diff
@@ -17,15 +17,16 @@
 
 (defmethod engine/apply-operation "core/combine"
   [{:keys [tenant-conn]} dataset-versions op-spec]
-  (let [namespace (engine/get-namespace op-spec)
```
Comment (tangrammer) [inline backend/src/akvo/lumen/lib/transformation/combine.clj:20]: we now derive `namespace` from transformation selected-column payload

File: backend/src/akvo/lumen/lib/transformation/delete_column.clj [hunk 7]
```diff
@@ -37,7 +37,8 @@
 (defmethod engine/apply-operation "core/delete-column"
   [{:keys [tenant-conn]} dataset-versions op-spec]
   (let [column-name (col-name op-spec)
-        namespace (engine/get-namespace op-spec)
```
Comment (tangrammer) [inline backend/src/akvo/lumen/lib/transformation/delete_column.clj:40]: we now derive `namespace` from transformation selected-column payload

File: backend/src/akvo/lumen/lib/transformation/derive.clj [hunk 8]
```diff
@@ -155,56 +155,56 @@
 
 (defmethod engine/apply-operation "core/derive"
   [{:keys [tenant-conn]} dataset-versions op-spec]
-  (let [namespace (engine/get-namespace op-spec)
+  (let [all-columns (engine/all-columns dataset-versions)
```
Comment (tangrammer) [inline backend/src/akvo/lumen/lib/transformation/derive.clj:158]: we now derive `namespace` from transformation selected-column payload

File: backend/src/akvo/lumen/lib/transformation/derive_category.clj [hunk 9]
```diff
@@ -49,17 +49,21 @@
 (defmethod engine/apply-operation "core/derive-category"
   [{:keys [tenant-conn]} dataset-versions op-spec]
   (let [op-spec               (walk/keywordize-keys op-spec)
-        namespace (engine/get-namespace op-spec)
```
Comment (tangrammer) [inline backend/src/akvo/lumen/lib/transformation/derive_category.clj:52]: we now derive `namespace` from transformation selected-column payload

File: backend/src/akvo/lumen/lib/transformation/engine.clj [hunk 10]
```diff
@@ -44,10 +44,14 @@
   false)
 
 (defn unify-transformation-history [dataset-versions]
-  (->> dataset-versions
-       (map :transformations)
-       (reduce into [])
-       (sort-by #(get % "created") #(compare (Instant/parse %1) (Instant/parse %2)))))
+  (if (= 1 (count dataset-versions))
```
Comment (tangrammer) [inline backend/src/akvo/lumen/lib/transformation/engine.clj:47]: backward compatibility with older flow or csv datasets

File: backend/src/akvo/lumen/lib/transformation/filter_column.clj [hunk 11]
```diff
@@ -12,7 +12,8 @@
   [{:keys [tenant-conn]} dataset-versions op-spec]
   (let [{expr "expression"
          column-name "columnName"} (engine/args op-spec)
-        namespace (engine/get-namespace op-spec)
```
Comment (tangrammer) [inline backend/src/akvo/lumen/lib/transformation/filter_column.clj:15]: we now derive `namespace` from transformation selected-column payload

File: backend/src/akvo/lumen/lib/transformation/geo.clj [hunk 12]
```diff
@@ -26,19 +26,19 @@
 
 (defmethod engine/apply-operation "core/generate-geopoints"
   [{:keys [tenant-conn]} dataset-versions op-spec]
-  (let [namespace (engine/get-namespace op-spec)
```
Comment (tangrammer) [inline backend/src/akvo/lumen/lib/transformation/geo.clj:29]: we now derive `namespace` from transformation selected-column payload

File: backend/src/akvo/lumen/lib/transformation/merge_datasets.clj [hunk 13]
```diff
@@ -139,18 +139,19 @@
 
 (defn apply-merge-operation
   [conn dataset-versions op-spec]
-  (let [namespace (engine/get-namespace op-spec)
```
Comment (tangrammer) [inline backend/src/akvo/lumen/lib/transformation/merge_datasets.clj:142]: we now derive `namespace` from transformation selected-column payload

File: backend/src/akvo/lumen/lib/transformation/multiple_column/caddisfly.clj [hunk 14]
```diff
@@ -34,7 +34,8 @@
   [{:keys [tenant-conn caddisfly] :as deps} dataset-versions op-spec]
   (jdbc/with-db-transaction [conn tenant-conn]
     (let [{:keys [onError op args]} op-spec
```
Comment (tangrammer) [inline backend/src/akvo/lumen/lib/transformation/multiple_column/caddisfly.clj:36]: we now derive `namespace` from transformation selected-column payload

File: backend/src/akvo/lumen/lib/transformation/multiple_column/geo_shape_features.clj [hunk 15]
```diff
@@ -29,7 +29,8 @@
   [{:keys [tenant-conn caddisfly] :as deps} dataset-versions op-spec]
   (jdbc/with-db-transaction [conn tenant-conn]
     (let [{:keys [onError op args]} op-spec
-          namespace (engine/get-namespace op-spec)
+          all-columns (engine/all-columns dataset-versions)
```
Comment (tangrammer) [inline backend/src/akvo/lumen/lib/transformation/multiple_column/geo_shape_features.clj:32]: we now derive `namespace` from transformation selected-column payload

File: backend/src/akvo/lumen/lib/transformation/rename_column.clj [hunk 16]
```diff
@@ -17,7 +17,8 @@
 
 (defmethod engine/apply-operation "core/rename-column"
   [{:keys [tenant-conn]} dataset-versions op-spec]
-  (let [namespace (engine/get-namespace op-spec)
```
Comment (tangrammer) [inline backend/src/akvo/lumen/lib/transformation/rename_column.clj:20]: we now derive `namespace` from transformation selected-column payload

File: backend/src/akvo/lumen/lib/transformation/reverse_geocode.clj [hunk 17]
```diff
@@ -30,11 +30,11 @@
 
 (defmethod engine/apply-operation "core/reverse-geocode"
   [{:keys [tenant-conn]} dataset-versions {:strs [args] :as op-spec}]
-  (let [namespace (engine/get-namespace op-spec)
```
Comment (tangrammer) [inline backend/src/akvo/lumen/lib/transformation/reverse_geocode.clj:33]: we now derive `namespace` from transformation selected-column payload

File: backend/src/akvo/lumen/lib/transformation/sort_column.clj [hunk 18]
```diff
@@ -21,7 +21,8 @@
   [{:keys [tenant-conn]} dataset-versions op-spec]
   (let [{column-name "columnName"
          sort-direction "sortDirection"} (engine/args op-spec)
-        namespace (engine/get-namespace op-spec)
```
Comment (tangrammer) [inline backend/src/akvo/lumen/lib/transformation/sort_column.clj:24]: we now derive `namespace` from transformation selected-column payload

File: backend/src/akvo/lumen/lib/transformation/split_column.clj [hunk 19]
```diff
@@ -81,22 +81,22 @@
   [{:keys [tenant-conn]} dataset-versions op-spec]
   (jdbc/with-db-transaction [tenant-conn tenant-conn]
     (let [{:keys [onError op args]} (walk/keywordize-keys op-spec)
-          namespace (engine/get-namespace op-spec)
```
Comment (tangrammer) [inline backend/src/akvo/lumen/lib/transformation/split_column.clj:84]: we now derive `namespace` from transformation selected-column payload

File: backend/src/akvo/lumen/lib/transformation/text.clj [hunk 20]
```diff
@@ -8,7 +8,8 @@
 (defn- transform
   [tenant-conn dataset-versions op-spec fn]
   (let [{column-name "columnName"} (engine/args op-spec)
-        namespace (engine/get-namespace op-spec)
```
Comment (tangrammer) [inline backend/src/akvo/lumen/lib/transformation/text.clj:11]: we now derive `namespace` from transformation selected-column payload

**Review by kardan — APPROVED** [review]
LGTM

**Pull request merged**